src/mqt/ionshuttler/linear/actions.py

```
@@ -0,0 +1,480 @@
+            UserWarning: If a physical gate has zero duration.
+        """
+        if not isinstance(self.virtual, bool):
+            msg = "virtual must be a boolean"
+            raise TypeError(msg)
+
+        duration = getattr(self, "duration", None)
+        if duration is None:
+            return
+        if isinstance(duration, bool) or not isinstance(duration, int):
+            msg = "action duration must be an integer"
+            raise ValueError(msg)  # ruff: ignore[type-check-without-type-error] - Model validation uses ValueError.
```

Ruff (0.16.1)

[warning] Prefer `TypeError` exception for invalid type

(TRY004)